### ISSUE TITLE
Flutter system bottom navigation bar startup animation

### ISSUE THREAD

**Issue by nt4f04uNd** [flutter/flutter#39614]
Basically, I want to build a whole black-themed flutter music player, so I want system bottom navigation bar to appear black too, but the bar blinks from white to black. 

Besides, when I tested it, I noticed that bar doesn't have startup "flight" animation, like almost any android application has nowadays, e.g. discord.

### See how this looks:
![How flutter starts up](https://i.imgur.com/lWzORu8.gif)

I thought this issue is related only for my app, but then I decided to test it out on other flutter applications:
- Flutter Gallery - doesn't have flight animation
- Google Ads - doesn't have flight animation and has terrible color blink effect after splash screen
- Alibaba - doesn't have flight animation

### And now see what I mean by word "flight"

![How any other android application starts up](https://i.imgur.com/cRRSHYv.gif)

The effect that you can see on first gif is achieved by specifying this code string in LaunchTheme
```xml
 <item name="android:navigationBarColor">#000</item>
```` 
Without this line the bar is white until the app starts and I switch to some route and go back to main screen, it's so weird.

I can't figure out, how to remove that annoying color delay and how to make startup animation the same as like discord has. I am new to native development so sorry if I am missing something.

[Github project](https://github.com/nt4f04uNd/sweyer)

[Original stackoverflow question](https://stackoverflow.com/questions/57733288/flutter-android-instant-system-bottom-navigation-color-and-fly-animation)

**Comment by markusaksli-nc** [comment]
Reproducible on the latest master `1.21.0-2.0.pre.55` with Android 9, but not Android 11. 
On an Android 11 (API 30) Pixel 3 emulator putting the style code in LaunchTheme or NormalTheme, both resulted in the correct behavior.
<img src="https://user-images.githubusercontent.com/68219924/87776190-0d93d900-c830-11ea-854b-31df765986f6.gif" height="400" />
On an SM G950F and Redmi Note 7, both using Android 9 (API 28) the style needs to be in NormalTheme to work and the issue reproduces. Adding it to both LaunchTheme and NormalTheme doesn't help. The issue isn't present on other apps with custom navbar color as the author described.
<img src="https://user-images.githubusercontent.com/68219924/87776217-184e6e00-c830-11ea-9c1d-9f6e9bbf8c07.gif" height="400" />


<details>
 <summary>flutter doctor -v</summary>

 ```
[√] Flutter (Channel master, 1.21.0-2.0.pre.55, on Microsoft Windows [Version 10.0.18362.900], locale et-EE)
    • Flutter version 1.21.0-2.0.pre.55 at C:\Development\flutter_master
    • Framework revision 1840b7121a (4 hours ago), 2020-07-16 23:15:23 -0700
    • Engine revision d3b81f19fc
    • Dart version 2.9.0 (build 2.9.0-21.0.dev 06183779d7)


[√] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at C:\Users\Isa\AppData\Local\Android\sdk
    • Platform android-30, build-tools 30.0.0
    • Java binary at: C:\Users\Isa\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\193.6626763\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Android Studio (version 4.0)
    • Android Studio at C:\Users\Isa\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\193.6626763
    • Flutter plugin version 47.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] Connected device (4 available)
    • SM G950F (mobile)       • ce12171c51cc001c03 • android-arm64  • Android 9 (API 28)
    • sdk gphone x86 (mobile) • emulator-5554      • android-x86    • Android 11 (API 30) (emulator)
    • Web Server (web)        • web-server         • web-javascript • Flutter Tools
    • Chrome (web)            • chrome             • web-javascript • Google Chrome 84.0.4147.89

• No issues found!
```
</details>

**Comment by nt4f04uNd** [comment]
@markusaksli-nc I guess you've missed that the black is the default color for the emulator. I didn't test, but I think this might be relatable to all Android versions

**Comment by markusaksli-nc** [comment]
You're right actually. With
```xml
<item name="android:navigationBarColor">#e7f471</item>
```
it's reproducible on both my Android 11 emulator and Android 9 physical on the latest master `1.24.0-4.0.pre.138`.
<details>
 <summary>flutter doctor -v</summary>

```
[√] Flutter (Channel master, 1.24.0-4.0.pre.138, on Microsoft Windows [Version 10.0.19041.572], locale en-US)
    • Flutter version 1.24.0-4.0.pre.138 at C:\Development\flutter_master
    • Framework revision 13896b3bd1 (4 hours ago), 2020-10-28 19:43:19 -0700
    • Engine revision 073263e39d
    • Dart version 2.11.0 (build 2.11.0-260.0.dev)

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at C:\Users\marku\AppData\Local\Android\sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: C:\Users\marku\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\201.6858069\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.6.5)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.6.30320.27
    • Windows 10 SDK version 10.0.18362.0

[√] Android Studio (version 4.1.0)
    • Android Studio at C:\Users\marku\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\201.6858069
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] IntelliJ IDEA Ultimate Edition (version 2020.3)
    • IntelliJ at C:\Users\marku\AppData\Local\JetBrains\Toolbox\apps\IDEA-U\ch-0\203.5251.39
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart

[√] Connected device (6 available)
    • SM G950F (mobile)           • ce12171c51cc001c03 • android-arm64  • Android 9 (API 28)
    • sdk gphone x86 arm (mobile) • emulator-5554      • android-x86    • Android 11 (API 30) (emulator)
    • Windows (desktop)           • windows            • windows-x64    • Microsoft Windows [Version 10.0.19041.572]
    • Web Server (web)            • web-server         • web-javascript • Flutter Tools
    • Chrome (web)                • chrome             • web-javascript • Google Chrome 86.0.4240.111
    • Edge (web)                  • edge               • web-javascript • Microsoft Edge 86.0.622.51

• No issues found!
```
</details>

**Comment by nt4f04uNd** [comment]
Also would like to note that this is actully crucial thing that is impossible to workaround - you either end up with logo jump, or with default system nav color https://github.com/flutter/flutter/issues/47634#issuecomment-583405878

**Comment by nt4f04uNd** [comment]
I tried to come up with some hack to this and actually, barely, but I did. Also, I would say that this is Android's rather than Flutter's blame we have such a bad API for manipulating the navbar/status bar colors.

There are two ways of working this around, which really depend on your needs.
- Are you okay with default system navigation bar?
**Give this up**
- Do you want to have just white or black splash background like most of the apps today (Instagram, Youtube Music)?
**Consider putting just the colored logo at the center of the splash and using `values` and `values-night` to specify light and dark backgrounds**
- Do you want to have a custom colored - not black and not white background for your application splash screen?
**If you answered yes to this question, read further**

### TL;DR
The thing is to make the navigation bar to be translucent during the whole application startup process, i.e. to apply it both to `LaunchTheme` and to `NormalTheme`, and then, when the splash screen disappears, make it not translucent.

But not all as simple as it sounds at the first glance. 

## What's wrong

Before I start - wondering why do we have to apply the translucent flag to both themes? Because, as I mentioned above, otherwise, you will see the logo jumps when the `NormalTheme` is applied.

> ... you either end up with logo jump, or with default system nav color [#47634 (comment)](https://github.com/flutter/flutter/issues/47634#issuecomment-583405878)

Now, let's talk about flaws of this approach.

1. The android disallows you to have [`SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR`](https://developer.android.com/reference/android/view/View#SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR), which is directly pointed in the documentation, so this leaves you with that during all the precense of the splash screen your navigation bar brightness will be 'dark', so the buttons will be light colored. And remember, we also apply this to `NormalTheme`, so you won't be able to change them for some time before the splash screen will fully disappear.

2. Translucent navbar disallows any color changes to it, making it fully transparent (actually, depends on system implementation, but in most cases; for example on my Redmi Note 5 it's fully transparent, in the Android R emulator it's sort of half-transparent). So the space occupied by the navbar will be filled with the `Scaffold`'s `backgroundColor`. If we have a custom navbar color we want to apply, we can fake it's presense by adding the `Overlay` of height `48` right at the bottom of the screen (that is assuming we have a portrait-only application). One more thing to worry about here - it to check for the bottom offset of the Android's `View` to not to show this fake overlay entry when we don't have the navbar at all and thus we don't need this hack.

3. Even applying these hacks, we are still at the point where we don't know when the animation of the splash screen will end, because it sometimes has different durations, which depends as I think on how accessible the application resources for the OS at the moment. I approximately hardcoded this, just used  the `const Duration` in my code. If someone know how do I exactly know when the splash screen will be 100% disappeared, that will be great 😊

## Video demonstration
I have the application I'm building where I would like to use this, so I can show you how this looks. I have two themes

- Light - https://youtu.be/u7smixn-JBQ
- Dark - https://youtu.be/3g4M5YJrXds

Looks awesome. But this solution is hard-coded and sometimes it still will show the logo jump, as I mentioned before.

It will be nothing but the amazing to see more options to customizes this in Flutter out of the box 🤗

## Code example

To be consistent, I want to show you directly on the code how I implemented this.

### Theme .xml configuration

```xml
<?xml version="1.0" encoding="utf-8"?>
<resources xmlns:tools="http://schemas.android.com/tools">

    <color name="main">#7c4dff</color>

    <!-- Show a splash screen on the activity. Automatically removed when
       Flutter draws its first frame -->
    <style name="LaunchTheme" parent="@android:style/Theme.Black.NoTitleBar">
        <item name="android:windowBackground">@drawable/launch_background</item>
        <item name="android:statusBarColor">@android:color/transparent</item>
        /// Also the status bar just to be sure
        <item name="android:windowTranslucentStatus">true</item>
        <item name="android:windowTranslucentNavigation">true</item>
    </style>

    <!-- The normal theme background only shows for a very brief moment after the splash screen disappears, 
        and during orientation change and Activity restoration.

        This Theme is only used starting with V2 of Flutter's Android embedding. -->
    <style name="NormalTheme" parent="@android:style/Theme.Black.NoTitleBar">
        <item name="android:statusBarColor">@android:color/transparent</item>
        /// Also the status bar just to be sure
        <item name="android:windowTranslucentStatus">true</item>
        <item name="android:windowTranslucentNavigation">true</item>
    </style>

</resources>
```

### The native part

There are two methods I implemented on the native side: to check if navbar height is 0, and another one to remove the translucency flags.

```java
// In some method channel handler...

// ...
case Constants.channels.general.METHOD_GET_NAV_BAR_HEIGHT:
    result.success(getNavigationBarSize(activity.getApplicationContext()).y);
    break;
case Constants.channels.general.METHOD_MAKE_NOT_TRANSLUCENT:
    if (activity != null) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            activity.getWindow().setBackgroundDrawable(new ColorDrawable(((Long) call.arguments).intValue()));
            activity.getWindow().clearFlags(FLAG_TRANSLUCENT_STATUS | FLAG_TRANSLUCENT_NAVIGATION);
        }
    }
    result.success(null);
    break;
// ...

// Methods taken from the https://stackoverflow.com/a/29609679/9710294

public static Point getNavigationBarSize(Context context) {
    Point appUsableSize = getAppUsableScreenSize(context);
    Point realScreenSize = getRealScreenSize(context);

    // navigation bar on the side
    if (appUsableSize.x < realScreenSize.x) {
        return new Point(realScreenSize.x - appUsableSize.x, appUsableSize.y);
    }

    // navigation bar at the bottom
    if (appUsableSize.y < realScreenSize.y) {
        return new Point(appUsableSize.x, realScreenSize.y - appUsableSize.y);
    }

    // navigation bar is not present
    return new Point();
}

public static Point getAppUsableScreenSize(Context context) {
    WindowManager windowManager = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
    Display display = windowManager.getDefaultDisplay();
    Point size = new Point();
    display.getSize(size);
    return size;
}

public static Point getRealScreenSize(Context context) {
    WindowManager windowManager = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
    Display display = windowManager.getDefaultDisplay();
    Point size = new Point();

    if (Build.VERSION.SDK_INT >= 17) {
        display.getRealSize(size);
    } else if (Build.VERSION.SDK_INT >= 14) {
        try {
            size.x = (Integer) Display.class.getMethod("getRawWidth").invoke(display);
            size.y = (Integer) Display.class.getMethod("getRawHeight").invoke(display);
        } catch (IllegalAccessException e) {
        } catch (InvocationTargetException e) {
        } catch (NoSuchMethodException e) {
        }
    }

    return size;
}
```

### The dart part

Firstly, the theme setup code.

```dart
class ThemeControl {

static bool _shouldShowNavbarPlaceholder;
/// Whether to show the fake overlay entry.
static bool get shouldShowNavbarPlaceholder => _shouldShowNavbarPlaceholder;

static Completer _initCompleter = Completer();
/// Future that completes after the navbar becomes translucent
static Future get waitTranslucent => _initCompleter?.future;

/// The function that's called in the Flutter main function
static Future<void> init() async {
    // A call to the Flutter channel to get the navbar height.
    final navbarHeight = await API.GeneralHandler.getNavbarHegiht();
    _shouldShowNavbarPlaceholder = navbarHeight > 0.0;
    
    final lightTheme = await Settings.lightThemeBool.getPref();
    _brightness = lightTheme ? Brightness.light : Brightness.dark;

    () async {
      // Some UI I want to apply after the splash screen
      var ui =
          Constants.AppSystemUIThemes.playerRouteScreen.autoBr(_brightness);
      var _sub = SystemUiControl.onUiChange.listen((newUi) {
        /// Update ui variable in case of that happens
        /// (for example user expands the player route before this functions completes).
        ui = newUi;
      });

      /// [SystemUiControl.setSystemUiOverlay] is just the function in my code base, it calls [SystemChrome.setSystemUIOverlayStyle] internally.

      /// At this moment the spalsh is still here.
      /// Keep app bar to be transparent and have light icons.
      SystemUiControl.setSystemUiOverlay(ui.copyWith(
        statusBarIconBrightness: Brightness.light,
        statusBarColor: Colors.transparent,
      ));

      await Future.delayed(const Duration(milliseconds: 700));

      if (lightTheme) {
        /// Approximately .7 of the animation is compelte - change [statusBarIconBrightness] dependent on theme,
        /// still have transparent status bar.
        SystemUiControl.setSystemUiOverlay(ui.copyWith(
          statusBarIconBrightness: Brightness.dark,
          statusBarColor: Colors.transparent,
        ));
      }

      await Future.delayed(const Duration(milliseconds: 100));

      /// The animation is ended completely.
      /// Remove translucent flag and default colors.
      SystemUiControl.setSystemUiOverlay(ui);
      await API.GeneralHandler.makeNotTranslucent(ui.systemNavigationBarColor);

      _initCompleter.complete();
      _sub.cancel();
    }();
  }
}
```

And later in the `main` function and the `App` widget

```dart

void main() async {
  // ...
  // Initializing our theme
  await ThemeControl.init();
}

class _AppState extends State<App> {
  @override
  void initState() {
    super.initState();

    /// The placeholder overlay entry

    if (ThemeControl.shouldShowNavbarPlaceholder) {
      WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
        final context = App.navigatorKey.currentContext;
        final screenWidth = MediaQuery.of(context).size.width;
        OverlayEntry overlayEntry;
        overlayEntry = OverlayEntry(
          builder: (context) => Align(
            alignment: Alignment.bottomCenter,
            child: FutureBuilder(
              future: ThemeControl.waitTranslucent,
              builder: (BuildContext context, AsyncSnapshot snapshot) {
                if (snapshot.connectionState == ConnectionState.done) {

                  /// Remove the entry as navbar is no longer translucent
                  WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
                    overlayEntry.remove();
                  });
                  return SizedBox.shrink();
                }

                /// Your custom placeholder.
                return StreamBuilder<SystemUiOverlayStyle>(
                  stream: SystemUiControl.onUiChange,
                  initialData: Constants
                      .AppSystemUIThemes.playerRouteScreen.autoWithoutContext,
                  builder: (context, snapshot) {
                    return Container(
                      height: Constants.navbarHeight,
                      width: screenWidth,
                      color: snapshot.data.systemNavigationBarColor,
                    );
                  },
                );
              },
            ),
          ),
        );

        App.navigatorKey.currentState.overlay.insert(overlayEntry);
      });
    }
  }
```

**Comment by Alioune05** [comment]
hello, is there any news for this issue?
I have the same problem but no viable solution that won't make my splash screen jumping up.

**Comment by nt4f04uNd** [comment]
no

**Comment by maheshmnj** [comment]
@nt4f04uNd I verified the issue on the latest master channel and I can no longer reproduce it.

https://user-images.githubusercontent.com/31410839/207964700-fb25f2cd-44db-48fa-9fde-575c4c6b3dae.mov


<details>
<summary>flutter doctor -v (mac)</summary>

```
[✓] Flutter (Channel master, 3.7.0-3.0.pre.33, on macOS 12.6 21G115 darwin-arm64, locale en-IN)
    • Flutter version 3.7.0-3.0.pre.33 on channel master 
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 5201856805 (38 minutes ago), 2022-12-05 18:27:21 -0800
    • Engine revision a309d239c4
    • Dart version 2.19.0 (build 2.19.0-463.0.dev)
    • DevTools version 2.20.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly
      to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc4)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0-rc4
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14A400
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 61.2.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.70.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.55.20221129

[✓] Connected device (3 available)
    • iPhone 12 Pro (mobile) • 026D5789-9E78-4AD5-B1B2-3F8D4E7F65E4 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-14-5 (simulator)
    • macOS (desktop)        • macos                                • darwin-arm64   • macOS 12.6 21G115 darwin-arm64
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 108.0.5359.94

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
```
[✓] Flutter (Channel stable, 3.3.9, on macOS 12.6 21G115 darwin-arm, locale en-IN)
    • Flutter version 3.3.9 on channel stable at /Users/mahesh/Development/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision b8f7f1f986 (24 hours ago), 2022-11-23 06:43:51 +0900
    • Engine revision 8f2221fbef
    • Dart version 2.18.5
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc4)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0-rc4
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14A400
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 61.2.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.70.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.53.20221101

[✓] Connected device (3 available)
    • iPhone 12 Pro (mobile) • 026D5789-9E78-4AD5-B1B2-3F8D4E7F65E4 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-14-5 (simulator)
    • macOS (desktop)        • macos                                • darwin-arm64   • macOS 12.6 21G115 darwin-arm
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 107.0.5304.110

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
</details>

So I believe that this issue is fixed, Closing. Incase you still encounter the issue, Please feel free to write in the comments and we will reopen it.

Thank you.
